File: logitrack_core/internal/model/fatigue_config.go
package model

import "time"

// FatigueConfig holds the global parameters for the driver fatigue detection
// model. Persisted in data/fatigue_config.json; editable by admin via
// /admin/fatigue-config.
type FatigueConfig struct {
	// RiskThresholds defines the composite-score colour bands.
	// green  → score ≤ GreenMax
	// amber  → GreenMax < score < RedMin
	// red    → score ≥ RedMin
	RiskThresholds RiskThresholds `json:"risk_thresholds"`

	// ── Per-test weights (0.0–1.0 each) ─────────────────────────────────────
	// The sum of weights for all ENABLED tests must equal 1.0.
	// Weights for disabled tests are ignored in the composite calculation.
	KSSWeight     float64 `json:"kss_weight"`
	VoiceWeight   float64 `json:"voice_weight"`
	TactileWeight float64 `json:"tactile_weight"`
	PVTWeight     float64 `json:"pvt_weight"`

	// ── Per-test enable flags ─────────────────────────────────────────────────
	// Disabled tests are excluded from the composite score; their weights are
	// redistributed proportionally among tests that have data available.
	KSSEnabled     bool `json:"kss_enabled"`
	VoiceEnabled   bool `json:"voice_enabled"`
	TactileEnabled bool `json:"tactile_enabled"`
	PVTEnabled     bool `json:"pvt_enabled"`

	// DailyResetHour defines the hour (0–23, ART timezone) at which a new
	// "logical day" begins for the fatigue check-in gate. Defaults to 0 (midnight).
	// Example: if set to 6, drivers who check in at 05:00 must check in again
	// after 06:00 even though the calendar date has not changed.
	DailyResetHour int `json:"daily_reset_hour"`

	// LastCheckinReset is set by the admin "reset check-ins" action (testing only).
	// GetTodayCheckin returns 404 when the stored check-in RecordedAt is before
	// this timestamp, forcing the driver to redo the gate.
	LastCheckinReset *time.Time `json:"last_checkin_reset,omitempty"`

	// BlockRouteOnRed controls whether a RED fatigue score automatically locks the
	// driver's route screen until a supervisor authorises resumption.
	// nil is treated as true for backward compatibility with configs saved before
	// this field existed. Supervisor notifications are sent regardless of this flag.
	BlockRouteOnRed *bool `json:"block_route_on_red,omitempty"`
}

// IsBlockRouteOnRed returns true unless BlockRouteOnRed is explicitly set to false.
// nil (field absent from old JSON) defaults to true.
func (c FatigueConfig) IsBlockRouteOnRed() bool {
	return c.BlockRouteOnRed == nil || *c.BlockRouteOnRed
}

// RiskThresholds defines the composite-score colour bands shown in the
// supervisor dashboard.
type RiskThresholds struct {
	GreenMax int `json:"green_max"` // upper bound for green (no-risk) zone
	RedMin   int `json:"red_min"`   // lower bound for red (high-risk) zone
}

func ptrBool(v bool) *bool { return &v }

func DefaultFatigueConfig() FatigueConfig {
	return FatigueConfig{
		RiskThresholds: RiskThresholds{
			GreenMax: 29,
			RedMin:   60,
		},
		KSSWeight:       0.15,
		VoiceWeight:     0.25,
		TactileWeight:   0.25,
		PVTWeight:       0.35,
		KSSEnabled:      true,
		VoiceEnabled:    true,
		TactileEnabled:  true,
		PVTEnabled:      true,
		BlockRouteOnRed: ptrBool(true),
	}
}

File: logitrack_web/src/api/fatigueConfig.ts
import axios from "axios";

const api = axios.create({
  baseURL: import.meta.env.VITE_API_URL ?? "http://localhost:8080/api/v1",
});

api.interceptors.request.use((config) => {
  const token = localStorage.getItem("token");
  if (token) config.headers.Authorization = `Bearer ${token}`;
  return config;
});

export interface RiskThresholds {
  green_max: number;
  red_min: number;
}

export interface FatigueConfig {
  risk_thresholds: RiskThresholds;

  // Per-test weights (0.0–1.0). Sum of enabled-test weights must equal 1.0.
  kss_weight:     number;
  voice_weight:   number;
  tactile_weight: number;
  pvt_weight:     number;

  // Per-test enable flags. Disabled tests are excluded from the composite score.
  kss_enabled:     boolean;
  voice_enabled:   boolean;
  tactile_enabled: boolean;
  pvt_enabled:     boolean;

  /** Hour (0–23, ART) at which a new logical day begins for the check-in gate. Default 0 = midnight. */
  daily_reset_hour: number;

  /** ISO timestamp; null/absent when no reset has been triggered yet. */
  last_checkin_reset?: string | null;
}

// ── Audit logs (US13) ────────────────────────────────────────────────────────

/** Immutable audit record. Naming mirrors MLConfig (created_at / created_by). */
export interface AuditLog {
  id: string;
  created_at: string;   // ISO 8601
  created_by: string;   // username of the actor (admin or driver)
  action: "UPDATE_FATIGUE_CONFIG" | "SUBMIT_CHECKIN" | "SKIP_CHECKIN";
  details: Record<string, unknown>; // JSON snapshot of what changed
}

export interface AuditLogsResponse {
  logs: AuditLog[];
  total: number;
}

export const fatigueConfigApi = {
  get: () =>
    api.get<FatigueConfig>("/admin/fatigue-config").then((r) => r.data),
  update: (cfg: FatigueConfig) =>
    api.put<FatigueConfig>("/admin/fatigue-config", cfg).then((r) => r.data),
  /** Forces every driver to redo (or skip) today's fatigue gate.
   *  Existing check-in data is preserved. */
  resetCheckins: () =>
    api.post<FatigueConfig>("/admin/fatigue-config/reset-checkins").then((r) => r.data),
  /** Returns all audit records newest-first. Strictly read-only (AC2). */
  getAuditLogs: () =>
    api.get<AuditLogsResponse>("/admin/audit-logs").then((r) => r.data),
};
